# PAUTA Laboratorio 2.2: Clasificación

Felipe Bravo-Marquez, Hernán Sarmiento, Gabriel Iturra-Bocaz, María José Zambrano

Mayo 2023

# Declaración de compromiso ético


Nosotros Diego Ignacio Faúndez Ortega y Javier Ignacio Kauer Lara, declaramos que realizamos de manera grupal los pasos de la presente actividad. También declaramos no incurrir en copia, ni compartir nuestras respuestas con otras personas ni con otros grupos. Por lo que, ratificamos que las respuestas son de nuestra propia confección y reflejan nuestro propio conocimiento.

# Instrucciones


1. Trabajen en equipos de dos personas. Salvo excepciones, no se corregirá entregas con menos de dos integrantes.

2. Modifique este archivo `.ipynb` agregando sus respuestas donde corresponda. 

3. El formato de entrega para esta actividad es un archivo **html**. Genere un archivo HTML usando jupyter (ver tutorial 2) y súbalo a U-Cursos. Basta con que **uno de los integrantes haga la entrega**. Si ambos hacen una entrega en U-Cursos, se revisará cualquiera de éstas.

# Estructura del laboratorio

Este laboratorio está conformado por preguntas teóricas de temas vistos en clases y preguntas prácticas (donde se requiere completar código) intercaladas con preguntas de interpretación de resultados y análisis. La parte práctica se divide en: 

1. Comparar clasificadores con ciertos *baselines* o clasificadores base.
2. Seleccionar hiperparámetros.
3. Trabajar con clases desbalanceadas.

# Teoría

#### 1. Explique cómo escoge un árbol de decisión el atributo raíz.

Respuesta: Para escoger el atributo raíz, se busca el atributo que maximize la "Information Gain". De esta forma se elige el atributo que produce los nodos más "puros".

#### 2. Explique la lógica de los clasificadores kNN.

Respuesta: El clasificador kNN usa los k puntos más cercanos para obtener la clasificación. Para esto se requiere que los records tengan una métrica de distancia para poder medir que tan cerca se encuentran entre ellos, basandose en la lógica que los puntos cercanos a un punto x, problablemente se van a parecer entre si.

#### 3. En clasificación, ¿cuándo es conveniente escalar los atributos?. Indique una técnica para escalar atributos.

Respuesta: Es conveniente escalar los atributos cuando algunos de estos empieza a dominar la métrica de la distacia. Una técnica para escalar atributos corresponde a la "normalización a media cero y vrianza unitaria".

#### 4. ¿Qué es overfitting y underfitting en machine learning?

Respuesta: Overfitting y underfitting en machine learning, el primero corresponde a que cuando se entrena con datos demasiados especificos no permite que el modelo tenga una funcionalidad general, ya que está entrenado con los datos poco variados y el segundo corresponde a que cuando el modelo está entrenado con datos muy escazos al aplicarlo en la vida real, no se logra se logra abstraer de manera amplia la categoria a clasificar.

# Parte 1: Comparar clasificadores

Una de las principales tareas en enfoques supervisados es evaluar diferentes clasificadores y encontrar el mejor rendimiento de alguno de ellos. Por ejemplo, si tenemos dos (o más) clasificadores y queremos compararlos entre sí, nos interesa responder: *¿Cuál de los clasificadores es el mejor?* 
Para responder esta pregunta, no existe una única solución. 

Lo que haremos a continuación será ejecutar diferentes clasificadores y compararlos en base a las métricas de Precision, Recall y F1-score.

## Pregunta 1.1  

Para realizar la evaluación de distintos clasificadores, vamos a crear la función `run_classifier()`, la cual evalúa un clasificador `clf` recibido como parámetro, un dataset `X,y` (features y target) y un número de tests llamado `num_test`. Esta función almacena y retorna los valores de precision, recall y f1-score en la variable `metrics` además de los resultados de predicción.

En base a lo anterior, incluya las sentencias que ajusten el modelo junto a su correspondiente predicción sobre los datos. **No use cross-validation ni tampoco el parámetro `random_state`.**


In [1]:
### COMPLETAR ESTE CÓDIGO

## run_classifier recibe un clasificador y un dataset (X, y)
## y opcionalmente la cantidad de resultados que se quiere obtener del clasificador

from sklearn.metrics import f1_score, recall_score, precision_score
from sklearn.model_selection import train_test_split
import numpy as np

def run_classifier(clf, X, y, num_tests=100):
    metrics = {'f1-score': [], 'precision': [], 'recall': []}
    
    for _ in range(num_tests):
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30)
        ### INICIO COMPLETAR ACÁ 
        
        clf.fit(X_train, y_train)
        predictions = clf.predict(X_test)
        
        #### TIP: en base a los set de entrenamiento, genere la variable 'predictions' 
        #### que contiene las predicciones del modelo
        

        
        ### FIN COMPLETAR ACÁ
        
        metrics['y_pred'] = predictions
        metrics['f1-score'].append(f1_score(y_test, predictions)) 
        metrics['recall'].append(recall_score(y_test, predictions))
        metrics['precision'].append(precision_score(y_test, predictions))
    return metrics

Luego de completar el código anterior, ejecute el siguiente bloque para comparar distintos clasificadores. 
Usaremos un **dataset de cáncer de mamas** para evaluar. La información del dataset se puede encontrar en el siguiente link: 
https://archive.ics.uci.edu/ml/datasets/Haberman%27s+Survival

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/giturra/Lab2.2/main/haberman.csv')

In [4]:
df.head()

Age of patient at time of operation   Patient's year of operation  \
0                                   30                            64   
1                                   30                            62   
2                                   30                            65   
3                                   31                            59   
4                                   31                            65   

    Number of positive axillary nodes detected   Survival status  
0                                            1                 1  
1                                            3                 1  
2                                            0                 1  
3                                            2                 1  
4                                            4                 1

In [5]:
# getting X and y
df.rename(columns={"age": "year", "nodes": "status"})
X = df.iloc[:, 0:3].to_numpy()
y = df.iloc[:, 3].to_numpy()

In [6]:
## ejecutar este código

from sklearn.datasets import load_breast_cancer
from sklearn.dummy import DummyClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB  # naive bayes
from sklearn.neighbors import KNeighborsClassifier #kNN
from sklearn.svm import SVC  # support vector machine


c0 = ("Base Dummy", DummyClassifier(strategy='stratified'))
c1 = ("Decision Tree", DecisionTreeClassifier(max_depth=5))
c2 = ("Gaussian Naive Bayes", GaussianNB())
c3 = ("KNN", KNeighborsClassifier(n_neighbors=10))
c4 = ("Support Vector Machines", SVC())

classifiers = [c0, c1, c2, c3, c4]

results = {}
for name, clf in classifiers:
    metrics = run_classifier(clf, X, y)   # hay que implementarla en el bloque anterior.
    results[name] = metrics
    print("----------------")
    print("Resultados para clasificador: ", name) 
    print("Precision promedio:", np.array(metrics['precision']).mean())
    print("Recall promedio:", np.array(metrics['recall']).mean())
    print("F1-score promedio:", np.array(metrics['f1-score']).mean())
    print("----------------\n\n")  

----------------
Resultados para clasificador:  Base Dummy
Precision promedio: 0.7233021338735947
Recall promedio: 0.7392097575533866
F1-score promedio: 0.729174400247855
----------------


----------------
Resultados para clasificador:  Decision Tree
Precision promedio: 0.7731345729455096
Recall promedio: 0.85397424158311
F1-score promedio: 0.8091407348257724
----------------


----------------
Resultados para clasificador:  Gaussian Naive Bayes
Precision promedio: 0.7632272013658422
Recall promedio: 0.9437324770715863
F1-score promedio: 0.8431220180375516
----------------


----------------
Resultados para clasificador:  KNN
Precision promedio: 0.7615639326876046
Recall promedio: 0.9419657007213275
F1-score promedio: 0.8410320171461031
----------------


----------------
Resultados para clasificador:  Support Vector Machines
Precision promedio: 0.7296090287002039
Recall promedio: 0.9863282652357301
F1-score promedio: 0.8378261233497308
----------------




## Pregunta 1.2

Analizando los resultados obtenidos de cada clasificador, y basándose en las métricas calculadas. ¿Cuál es el mejor clasificador? ¿Qué métricas observó para tomar esa decisión y por qué? considerando el problema que aborda. Fundamente su respuesta.

(Considere *1* como clase positiva, y *2* como clase negativa.)

Respuesta: El mejor clasificador a elegir corresponde a el "Support Vector Machines", esto debido a que corresponde al valor de recall más alto, lo cual significa que tiene la menor cantidad de falsos negativos. Aplicando esto al problema, significa que este clasificador tiene menor cantidad de canceres que pasan inadvertidos. Siguiendo esta lógica, a pesar que el modelo no tenga la mayor precisión, es preferible tener una mayor cantidad de personas sanas mal diagnosticada que al revés.


## Pregunta 1.3

Analice el recall de la clase malignant ¿Cuál sería la razón de maximizar está métrica?

Respuesta: La razón para máximizar esta métrica es para minimizar los falsos diagnósticos de cancer. Aunque de todas formas es preferible hacer un falso diagnóstico de cancer que un falso diagnósitco de salud. 

# Parte 2: Seleccionar hiperparámetros

Los hiperparámetros son parámetros que no se aprenden directamente dentro de los estimadores. En scikit-learn se pasan como argumentos al constructor de las clases, por ejemplo, cuál kernel usar para Support Vector Classifier, o qué criterio para Decision Tree, etc. Es posible y recomendable buscar en el espacio de hiperparámetros la mejor alternativa.

Tenga en cuenta que es común que un pequeño subconjunto de esos parámetros pueda tener un gran impacto en el rendimiento predictivo o de cálculo del modelo, mientras que otros pueden dejar sus valores predeterminados. Se recomienda leer la documentación de la clase de estimador para obtener una mejor comprensión de su comportamiento esperado.

**Dataset:** En esta y la siguiente parte del laboratorio utilizaremos el dataset **"ML Classification: Predicting 5-Year Career Longevity for NBA Rookies"** de data.world (https://data.world/ssaudz/ml-classification-predicting-5-year-career-longevity-for-nb). Este dataset contiene estadísticas de los novatos en la NBA y busca predecir si un jugador podrá durar 5 años en la liga. La columna objetivo es *TARGET_5Yrs*. Esta es una versión preprocesada del dataset original (después de eliminar registros duplicados por nombre del jugador, anonimizar los datos, y eliminar los registros con valores nulos).

In [7]:
import pandas as pd
data = pd.read_csv('https://raw.githubusercontent.com/giturra/Lab2.2/main/longevity.csv')
data.shape

(1284, 20)

In [8]:
data.head(2)

GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  FTM  FTA   FT%  OREB  \
0  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  1.6  2.3  69.9   0.7   
1  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  2.6  3.4  76.5   0.5   

   DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0   3.4  4.1  1.9  0.4  0.4  1.3            0  
1   2.0  2.4  3.7  1.1  0.5  1.6            0

In [9]:
#separando atributos predictivos (X) del atributo objetivo (y)
X = data.iloc[:,:-1].values
y = data['TARGET_5Yrs'].values

#dividiendo los datos de entrenamiento y validación
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30,
                                                    random_state=15, stratify=y)

## GridSearchCV

Una alternativa para seleccionar hiperparámetros es GridSearchCV, la cual considera exhaustivamente todas las combinaciones de parámetros. GridSearchCV recibe un `estimador`, `param_grid` (un diccionario o una lista de diccionarios con los nombres de los parámetros a probar como keys y una lista de los valores a probar), `scoring` una o varias funciones de puntuación (score) para evaluar cada combinación de parámetros (opciones válidas: https://scikit-learn.org/stable/modules/model_evaluation.html), y `cv` una extrategia para hacer validación cruzada.

El siguiente código muestra cómo seleccionar el número de vecinos y qué pesos otorgar a los vecinos en un clasificador KNN.

In [18]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report


#Configure tuned_parameters
tuned_parameters = {'n_neighbors': [1, 3, 5, 10], 
                    'weights': ['uniform','distance']}

#set scoring metric
score = 'precision' 

#Construir el clf con GridSearch
clf = GridSearchCV(KNeighborsClassifier(), 
                   param_grid=tuned_parameters, 
                   cv=5,
                   scoring=score)

#Entrenar clf
clf.fit(X_train, y_train)

print("Mejor combinación de parámetros:")
print(clf.best_params_)
 
y_pred = clf.predict(X_test)

print(classification_report(y_test, y_pred))

Mejor combinación de parámetros:
{'n_neighbors': 10, 'weights': 'uniform'}
              precision    recall  f1-score   support

           0       0.64      0.60      0.62       145
           1       0.77      0.80      0.78       241

    accuracy                           0.73       386
   macro avg       0.71      0.70      0.70       386
weighted avg       0.72      0.73      0.72       386



## Pregunta 2.1

*  a) Utilizando los datos del bloque anterior (NBA_career_longevity.csv), realice este mismo proceso para un clasificador `DecisionTree` y los parametros `criterion=['gini','entropy']`, `max_depth=[3,5,7,10]` y tomando como `scoring` metric `'f1'`. Use `cv=10`.
*  b) ¿Qué puede decir de los resultados, con cuáles parámetros los obtuvo (revise que su respuesta concuerde con los resultados que imprime)? ¿Cuál considera que es la principal ventaja de aplicar GridSearchCV? ¿Considera que es necesario seguir explorando los parámetros?

In [19]:
## RESPUESTA A PREGUNTA 2.1 a)

## COMPLETE ACÁ
criterion=['gini','entropy']
max_depth=[3,5,7,10]

tuned_parameters = {'criterion': criterion, 'max_depth': max_depth}
score = 'f1'

clf = GridSearchCV(DecisionTreeClassifier(), param_grid=tuned_parameters, cv=10, scoring=score)
clf.fit(X_train, y_train)

print("Mejor combinación de parámetros:")
print(clf.best_params_)
 
y_pred = clf.predict(X_test)
print(classification_report(y_test, y_pred))

Mejor combinación de parámetros:
{'criterion': 'gini', 'max_depth': 3}
              precision    recall  f1-score   support

           0       0.62      0.59      0.61       145
           1       0.76      0.78      0.77       241

    accuracy                           0.71       386
   macro avg       0.69      0.69      0.69       386
weighted avg       0.71      0.71      0.71       386



Respuesta b: Los parametros que se usaron son, 'gini' y un max_depth de 3. La principal ventaja de GridSearchCV, que se ahorra tiempo al pasar muchos parametros y elegir los mejores. Si es necesario seguir explorando los parámetros, porque podrían haber mejores resultados para las métricas obtenidas (precision, recall, y f1-score).

# Parte 3: Trabajar con clases desbalanceadas

Al explorar el dataset anterior, se nota un desbalance importante (38%-62%). Para mejorar el rendimiento de un clasificador sobre clases desbalanceadas existen varias técnicas. En esta parte, veremos cómo tratar con este problema usando (sub/over) sampling de las clases.

(*Nota: Para ejecutar el siguiente bloque es necesaria la librería `pandas` que viene incluida en Anaconda.*)

Note el desbalance de las clases ejecutando el siguiente código:

In [20]:
print("Distribucion de clases original")
data['TARGET_5Yrs'].value_counts()

Distribucion de clases original


1    802
0    482
Name: TARGET_5Yrs, dtype: int64

Antes de hacer algo para tratar el desbalance entre las clases primero debemos dividir en train-test. Como ya hicimos la partición de train y test, vamos a explorarla a continuación.

In [21]:
#Dividmos igual que arriba. 
#Para facilitar el balance manual, X contendrá el dataset completo, pero luego eliminaremos de este el atributo objetivo.

data_train, data_test, ytrain, ytest = train_test_split(data, data['TARGET_5Yrs'], test_size=.30,
                                                random_state=15, stratify=data['TARGET_5Yrs'])

print("Cantidad de instancias por clase en el train:")
print("Clase 1: " + str((ytrain==1).sum()))
print("Clase 0: " + str((ytrain==0).sum()))

Cantidad de instancias por clase en el train:
Clase 1: 561
Clase 0: 337


Aplicaremos **oversampling** y **subsampling** al train para que queden balanceados. Ejecute el siguiente código y note ahora que las clases están balanceadas. 

In [22]:
print("Distribución de clases usando (over/sub) sampling: \n")
data_train = data_train.reset_index(drop=True)

# oversampling sobre la clase 0
idx = np.random.choice(data_train[data_train['TARGET_5Yrs'] == 0].index, size=224)
data_oversampled = pd.concat([data_train, data_train.iloc[idx]])
print("Data oversampled on class '0'")
print(data_oversampled['TARGET_5Yrs'].value_counts())
print()


# subsampling sobre la clase 1
idx = np.random.choice(data_train.loc[data_train.TARGET_5Yrs == 1].index, size=224, replace=False)
data_subsampled = data_train.drop(data_train.iloc[idx].index)
print("Data subsampled on class '1'")
print(data_subsampled['TARGET_5Yrs'].value_counts())

Distribución de clases usando (over/sub) sampling: 

Data oversampled on class '0'
1    561
0    561
Name: TARGET_5Yrs, dtype: int64

Data subsampled on class '1'
1    337
0    337
Name: TARGET_5Yrs, dtype: int64


**Nota:** *Librerías como `imblearn` son muy útiles para balancear los datos.*

## Pregunta 3. 1

¿Por qué aplicar subsampling/oversampling de las clases sobre el conjunto de entrenamiento en lugar de aplicarlo sobre el dataset completo? Argumente su respuesta.

**Respuesta:** La razón por la cual se aplica subsampling/oversampling porque de esta forma se tiene un conjunto de entrenamiento más balacneado, y por ende el modelo sea más representativo una vez aplicado. De esta forma se evita errores de underfitting/overfitting, y además permite evaluar el rendimiento del modelo directamente con la distribución real de datos.

In [24]:
## ejecutar este código para preparar los datos
from sklearn.metrics import classification_report

# Preparando los data frames para ser compatibles con sklearn

# datos test (mismo para todos los conjuntos de entrenamiento)
X_test = data_test[data_train.columns[:-1]] # todo hasta la penultima columna
y_test = data_test[data_train.columns[-1]]  # la última columna

# datos entrenamiento "originales"
X_orig = data_train[data_train.columns[:-1]] 
y_orig = data_train[data_train.columns[-1]] 

# datos entrenamiento "oversampleados" 
X_over = data_oversampled[data_train.columns[:-1]]
y_over = data_oversampled[data_train.columns[-1]]

# datos entrenamiento "subsampleados"
X_subs = data_subsampled[data_train.columns[:-1]]
y_subs = data_subsampled[data_train.columns[-1]]


## Pregunta 3.2

Complete el código necesario para entrenar un clasificador DecisionTree en cada uno de los tres casos (**original**, con **oversampling** y con **subsampling**) y luego compare los resultados sobre el conjunto de test (este es el mismo para los tres casos) obtenido con train_test_split sobre los datos originales. Muestre Precision, Recall y F1-score.

Emplee como datos de entrada lo del bloque anterior. 

In [36]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

## Pasos:
##  - instanciar el clasificador con DecisionTreeClassifier()
##  - entrenar con fit()
##  - hacer las predicciones
##  - mostrar precision, recall y f1-score con classification report.

print("ORIGINAL::::::::::")
clf_orig = DecisionTreeClassifier()
clf_orig.fit(X_orig, y_orig)
pred_orig = clf_orig.predict(X_test)
print(classification_report(y_test, pred_orig))

print("Oversampling::::::")
clf_over=DecisionTreeClassifier()
clf_over.fit(X_over, y_over)
pred_over = clf_over.predict(X_test)
print(classification_report(y_test, pred_over))

print("Subsampling:::::::")
clf_subs=DecisionTreeClassifier()
clf_subs.fit(X_subs, y_subs)
pred_subs = clf_subs.predict(X_test)
print(classification_report(y_test, pred_subs))


ORIGINAL::::::::::
              precision    recall  f1-score   support

           0       0.53      0.49      0.51       145
           1       0.71      0.74      0.72       241

    accuracy                           0.65       386
   macro avg       0.62      0.62      0.62       386
weighted avg       0.64      0.65      0.64       386

Oversampling::::::
              precision    recall  f1-score   support

           0       0.53      0.52      0.52       145
           1       0.71      0.73      0.72       241

    accuracy                           0.65       386
   macro avg       0.62      0.62      0.62       386
weighted avg       0.65      0.65      0.65       386

Subsampling:::::::
              precision    recall  f1-score   support

           0       0.50      0.57      0.53       145
           1       0.72      0.66      0.69       241

    accuracy                           0.63       386
   macro avg       0.61      0.61      0.61       386
weighted avg     

## Pregunta 3.3

- Observe los resultados obtenidos por clase con cada conjunto de entrenamiento, ¿se puede observar alguna diferencia importante? 
- Indique una desventaja de usar oversampling y una desventaja de usar subsampling en clasificación.

Respuesta: Si se puede observar una diferencia importante entre los resultados obtenidos. Por ejemplo, existe una variación considerable del recall entre los 3 modelos, y una diferencia menor en las otras métricas. Con respecto a las desventajas, en subsampling corresponde a que puede haber perdida de información que afecte el modelo, y con respecto a oversampling ocurre que se genera un sesgo hacia la clase con menor cantidad de datos (aunque esto igual puede ocurrir al usar subsampling).

## Pregunta 3.4

Compare los resultados del caso **ORIGINAL** (donde el clasificador usa los parámetros por defecto DecisionTreeClassifier()) versus el resultado de la pregunta 2.1 donde usa los mejores parámetros con GridSearchCV. ¿Qué opina de los resultados?

Respuesta: Se puede ver que las métricas para la clase 0 aumentan de manera significativa (unas varias centésimas) del caso ORIGINAL al caso de la preguntas 2.1, mientras que en la clase 1 aumenta no tan drásticamente (unas cuantas centésimas). Por lo que en general se puede apreciar que aplicar estas técnicas de hiperparámetros contribuyen a un mejor modelo.